apps/emqx/src/emqx_trace/emqx_trace.erl

```diff
@@ -688,22 +690,33 @@ stop_trace(#{id := HandlerID, dst := Basename, filter := {Type, Filter}} = Handl
 clean_stale_trace_files() ->
     TraceDir = trace_dir(),
     case file:list_dir(TraceDir) of
-        {ok, AllFiles} when AllFiles =/= ["zip"] ->
-            KeepFiles = [
+        {ok, []} ->
+            ok;
+        {ok, ["zip"]} ->
+            ok;
+        {ok, Files} ->
+            %% NOTE: Cleaning both stale and foreign files.
+            TraceFiles = [
                 log_filename(Name, StartAt)
              || #?TRACE{name = Name, start_at = StartAt} <- list_traces()
             ],
-            case AllFiles -- ["zip" | KeepFiles] of
-                [] ->
-                    ok;
-                DeleteFiles ->
-                    DelFun = fun(F) -> file:delete(filename:join(TraceDir, F)) end,
-                    lists:foreach(DelFun, DeleteFiles)
-            end;
-        _ ->
+            StaleFiles = [Filename || Filename <- Files, trace_file_is_stale(Filename, TraceFiles)],
+            lists:foreach(
+                fun(Filename) ->
+                    ?tp(debug, "trace_cleaning_stale_file", #{filename => Filename}),
+                    file:delete(filename:join(TraceDir, Filename))
+                end,
+                StaleFiles
+            );
+        _Error ->
             ok
     end.
 
+trace_file_is_stale("zip", _TraceFiles) ->
+    false;
```

Comment:
maybe not the scope of this pr: zip dir is only cleaned up in the `terminate` callback, if the system is never gracefully shutdown. there could be files never gets deleted.

Reply:
Good point. Looks like a [minor change](https://github.com/emqx/emqx/pull/15986/commits/da0fa13740bea9bc89d961e91bcd8498e03e1333), now included in this PR.